ruby/red-arrow/ext/arrow/raw-records.cpp

```diff
@@ -144,6 +144,118 @@ namespace red_arrow {
       // The number of columns.
       const int n_columns_;
     };
+
+    class RawRecordsProducer : private Converter, public arrow::ArrayVisitor {
+    public:
+      explicit RawRecordsProducer()
+        : Converter(),
+          record_(Qnil) {
+      }
+
+      void produce(const arrow::RecordBatch& record_batch) {
+        rb::protect([&] {
+          auto n_columns = record_batch.num_columns();
+          const auto n_rows = record_batch.num_rows();
+          for (int64_t i = 0; i < n_rows; ++i) {
+            record_ = rb_ary_new_capa(n_columns);
+            row_offset_ = i;
+            for (int i = 0; i < n_columns; ++i) {
+              const auto array = record_batch.column(i).get();
+              column_index_ = i;
+              check_status(array->Accept(this),
+                           "[record-batch][each-raw-record]");
+            }
+            rb_yield(record_);
+          }
+          return Qnil;
+        });
+      }
+
+      void produce(const arrow::Table& table) {
+        rb::protect([&] {
+          int n_columns = table.num_columns();
+          const auto& base_column = table.column(0);
+          for (int i = 0; i < base_column->num_chunks(); ++i) {
+            int chunk_length = base_column->chunk(i)->length();
+            for (int j = 0; j < chunk_length; ++j) {
+              row_offset_ = j;
+              record_ = rb_ary_new_capa(n_columns);
+              for (int k = 0; k < n_columns; ++k) {
+                column_index_ = k;
+                const auto& array = table.column(k)->chunk(i);
+                check_status(array->Accept(this),
+                             "[table][each-raw-record]");
+              }
+              rb_yield(record_);
+            }
+          }
```

Comment:
We can't assume that all columns have the same chunk layout.
For example, the first column has `[[1], [2, 3], [4]]` chunks but the second column may have `[[true, false], [false, true]]`.

How about keeping row offset for each column?

```suggestion
          auto n_columns = table.num_columns();
          auto n_rows = table.num_rows();
          std::vector<int> chunk_indexes(n_columns);
          std::vector<int64_t> row_offsets(n_columns);
          for (int64_t i_row = 0; i_row < n_rows; ++i_row) {
            record_ = rb_ary_new_capa(n_columns);
            for (int i_column = 0; i_column < n_columns; ++i_column) {
              column_index_ = i_column;
              const auto chunked_array = table.column(i_column).get();
              auto& chunk_index = chunk_indexes[i_column];
              auto& row_offset = row_offsets[i_column];
              auto array = chunked_array->chunk(chunk_index).get();
              while (array->length() == row_offset) {
                ++chunk_index;
                row_offset = 0;
                array = chunked_array->chunk(chunk_index).get();
              }
              row_offset_ = row_offset;
              check_status(array->Accept(this),
                           "[table][each-raw-record]");
              ++row_offset;
            }
            rb_yield(record_);
          }
```

Reply:
> We can't assume that all columns have the same chunk layout.

Thank you so much. I misunderstood this point🙏